pkg/services/accesscontrol/resourcepermissions/api_adapter.go

```diff
@@ -162,3 +165,223 @@ func getMapKeys(m map[string][]string) []string {
+		if perm.Permission == "" {
+			continue
+		}
+
+		kind := a.getPermissionKind(perm)
+		name := a.getPermissionName(ctx, perm)
+
+		if name == "" {
+			continue
+		}
+
+		k8sPermissions = append(k8sPermissions, map[string]interface{}{
+			"kind": kind,
+			"name": name,
+			"verb": perm.Permission,
+		})
+	}
+
+	if len(k8sPermissions) == 0 {
+		if existing != nil {
+			err = resourcePermResource.Delete(ctx, resourcePermName, metav1.DeleteOptions{})
+			if err != nil && !k8serrors.IsNotFound(err) {
+				return fmt.Errorf("failed to delete resource permission in k8s: %w", err)
+			}
+		}
+		return nil
+	}
+
+	resourcePerm := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": iamv0.ResourcePermissionInfo.GroupVersion().String(),
+			"kind":       iamv0.ResourcePermissionInfo.TypeMeta().Kind,
+			"metadata": map[string]interface{}{
+				"name":      resourcePermName,
+				"namespace": namespace,
+			},
+			"spec": map[string]interface{}{
+				"resource": map[string]interface{}{
+					"apiGroup": a.getAPIGroup(),
+					"resource": a.service.options.Resource,
+					"name":     resourceID,
+				},
+				"permissions": k8sPermissions,
+			},
+		},
+	}
+
+	if err == nil && existing != nil {
+		resourcePerm.SetResourceVersion(existing.GetResourceVersion())
+		_, err = resourcePermResource.Update(ctx, resourcePerm, metav1.UpdateOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to update resource permission in k8s: %w", err)
+		}
+	} else {
+		_, err = resourcePermResource.Create(ctx, resourcePerm, metav1.CreateOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to create resource permission in k8s: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (a *api) setUserPermissionToK8s(ctx context.Context, namespace string, resourceID string, userID int64, permission string) error {
+	userDetails, err := a.service.userService.GetByID(ctx, &user.GetUserByIDQuery{ID: userID})
+	if err != nil {
+		return fmt.Errorf("failed to get user details: %w", err)
+	}
+
+	return a.setSinglePermissionToK8s(ctx, namespace, resourceID, string(iamv0.ResourcePermissionSpecPermissionKindUser), userDetails.UID, permission)
+}
+
+func (a *api) setBuiltInRolePermissionToK8s(ctx context.Context, namespace string, resourceID string, builtInRole string, permission string) error {
+	return a.setSinglePermissionToK8s(ctx, namespace, resourceID, string(iamv0.ResourcePermissionSpecPermissionKindBasicRole), builtInRole, permission)
+}
+
+func (a *api) setSinglePermissionToK8s(ctx context.Context, namespace string, resourceID string, kind string, name string, permission string) error {
+	dynamicClient, err := a.getDynamicClient(ctx)
+	if err != nil {
+		return err
+	}
+
+	resourcePermName := a.buildResourcePermissionName(resourceID)
+	resourcePermResource := dynamicClient.Resource(iamv0.ResourcePermissionInfo.GroupVersionResource()).Namespace(namespace)
+
+	existing, err := resourcePermResource.Get(ctx, resourcePermName, metav1.GetOptions{})
```

Comment:
We should strongly type the resource permission object rather than using `unstructured` 🙏